### ISSUE TITLE
Performance Issues on Mavericks

### ISSUE THREAD

**Issue by hhliu0** [CouleeApps/MenuBarFilter#3]
High CPU usage (~10-15%), high number of idle wake ups
Activates dedicated GPU, fan speed increases to ~5000 rpm
All in all, too high energy and processor usage for an idle background app to be usable. Plus fan is very loud!
(Is this caused by incorrect Build Settings in XCode? Architectures is set to 32-bit Intel and Deployment Target set to OS X 10.4 in the .xcodeproj as downloaded. However, building as is vs changing the two settings to 64-bit and OS X 10.9 respectively seems to make no difference.)
![screen shot 2014-02-07 at 11 11 19 pm](https://f.cloud.github.com/assets/1537224/2111094/c9444828-900c-11e3-9c69-070f329b1718.png)
![screen shot 2014-02-07 at 11 12 05 pm](https://f.cloud.github.com/assets/1537224/2111093/c943d94c-900c-11e3-9d43-b9e3de3b7fe8.png)
![screen shot 2014-02-07 at 11 13 12 pm](https://f.cloud.github.com/assets/1537224/2111092/c941e2b8-900c-11e3-8ca7-6cb406f4446d.png)

**Comment by CouleeApps** [comment]
Looked into it, seems to be mostly the filtering done on the window. Will attempt to improve performance when I have more time.

**Comment by hhliu0** [comment]
Thanks, so it's not an issue with my build settings then? Does it make any difference if I change the 32-bit Intel and Deployment Target OS X 10.4 to 64-bit and OS X 10.9? Sorry, don't know too much about Objective-C and Cocoa.

**Comment by CouleeApps** [comment]
I don't think that would make much of a difference. The problem wasn't with the build settings– it was with the render code. I'm currently working on it right now.

**Comment by CouleeApps** [comment]
Ok whatever I did, I don't think it changed much. The problem always seems to be in the filtering of the image, rather than in the capturing of the image. If only Apple provided faster CGImageRef methods....

**Comment by hhliu0** [comment]
Thanks for looking into it, so my understanding is CGSAddWindowFilter() doesn't work anymore in Mavericks, that's why you need to use the CGImageRef method you mentioned? You've probably seen these threads: http://stackoverflow.com/questions/19575642/how-to-use-cifilter-on-nswindow-in-osx-10-9-mavericks http://stackoverflow.com/questions/14163788/how-does-on-screen-color-inversion-work-in-os-x, but if not perhaps they may be of some help to you.

Again, thanks for your time trying to solve this issue.

**Comment by hhliu0** [comment]
Oh and I noticed you updated the README file, but the information on the two hidden preferences is still in there. I seem to remember reading a comment in the code that those two options don't work anymore? Or are supposed to still work?

**Comment by CouleeApps** [comment]
I have the first one starred, actually. Yes, those no longer work as of 10.9 (they were a private API anyway, so Apple wasn't obligated to keep them). That is essentially why I have to use a CGImageRef. CGSAddWindowFilter() seems to now only work for blur filters (at least, as far as I have tested), and the inversion / gamma filters are completely broken.

**Comment by hhliu0** [comment]
I've tried your latest version, it is already much, much better than I expected. For one, it doesn't make the dedicated GPU kick in. CPU is around 6-7%, fans still spin above normal. But a huge improvement, so thanks! I'll try it out for a few days and monitor the performance in different situations, see if impact on battery life is too significant for it to be usable.

Good luck, hopefully you'll find some way of increasing the performance even more! You mentioned in the commit notes that some filters broke when you changed the code to increase performance, but I personally think the darker blue doesn't look too bad at all.

**Comment by hhliu0** [comment]
Actually, for some reason the CPU is now being used up by the WindowServer process instead of MenuBarFilter itself. As soon as MenuBarFilter is launched, WindowServer instantly jumps to ~20-30% CPU usage, and stays there. That's even higher than how much MenuBarFilter used originally. It goes back to 2-3% as soon as I kill the MenuBarFilter process, so I'm fairly sure that's what's causing it. Could you please look into it? I understand there might not be a lot you can do, so I really appreciate your time and effort.

**Comment by CouleeApps** [comment]
Problem with this is that it's _really_ slow to take a screenshot. I've been looking into quite a few methods and haven't found much else than this.

**Comment by CouleeApps** [comment]
Sped up performance by not capturing as often. Is now slower, but it's hardly noticeable. CPU usage (for me) is down to ~10% WindowServer, 2% MenuBarFilter. Let me know if you experience otherwise.

**Comment by hhliu0** [comment]
Awesome, it's way better now, thanks! I'm seeing very similar CPU usage: ~10-15% for WindowServer and 1-2% for MenuBarFilter. Thanks for your hard work, I think the performance is good enough for everyday usage now. But I guess we can still keep the issue open.

**Comment by hhliu0** [comment]
So, just a more general thing, if you'll pardon my ignorance: could you please briefly explain to me what private APIs are exactly and what reason Apple might have had to getting rid of the ones that used to work really well? Is it just arbitrary that they decide they'll no longer have those methods? And is there any possibility of requesting them (perhaps through the Apple developer bug reporting page?) to re-implement those APIs? Again, sorry if these questions rather simplistic, I'm just hoping to learn a bit more about programming.Thanks!

**Comment by CouleeApps** [comment]
Private APIs are undocumented "unofficial" APIs, in that they may exist in the binaries (like CGSAddWindowFilter), but are not mentioned in any public header files. As Apple has never publicly made known these APIs, they assume that developers aren't using them. Many times these functions are changed or removed completely, but because they are private, there is no clear "reason." Generally, asking about private APIs will turn away just about any Apple support person, as they technically "don't exist." Also, using private APIs like this is a sure-fire way of getting your app rejected from the App Store.

Whatever the case is, we're out of luck in terms of getting CGSAddWindowFilter back. There is the slightest chance that it may return in a future update, but that chance is so small it's not worth hoping for.

**Comment by hhliu0** [comment]
Ah I see, thanks for your clear and simple response!
